docs/modules/candid-guide/index.adoc

```diff
@@ -0,0 +1,1098 @@
+};
+service address_book : {
+  set_address: (name : text, address) -> ();
+  get_address: (name : text) -> (opt address) query;
+}
+....
+
+These type definitions merely name an _existing_ type, they do not define a new type. It does not matter whether you use `address` in the function signature, or write out the records. In orther words, Candid uses _structural_ typing.
+
+In the last example, we also see the use of the `query` annotation. This Internet-Computer specific annotation indicates that the method `get_address` can be invoked using a query call (TODO: reference general documentation about query calls here)
+
+=== Encoding and decoding
+
+The point of Candid is to allow seamless invocation of service methods, passing arguments encoded to a binary format and transferred by an underlying transporation method (such as messages into or within the Internet Computer), and decoded on the other side.
+
+As a Candid user, you do not have to worry about the details of this binary format. If you plan to _implemnet_ Candid yourself (e.g. for a new host language), you can consult the Candid specification for details. However, some aspects of the format are worth knowing:
+
+* The Candid binary format starts with `DIDL…` (or, in hex, `4449444c…`). If you see this in some low-level log output, you are very likely observing a Candid-encoded value.
+
+* The Candid binary format always encodes _sequences_ of values, because methods parameters and results are sequences of types.
+
+* The binary format is quite compact. A `(vec nat64)` with 125000 entries takes 1000007 bytes.
+
+* The binary is self-describing, and includes a (condended) description of type of the values therein. This allows the receiving side to detect if a message was sent at a different, incompatible type.
+
+* As long as the sending side serializes the arguments at the same type that the receiving side expects them, deserialization will succeed.
+
+[#upgrades]
+=== Service upgrades
+
+Services evolve over time: They gain new methods, existing methods return more data, or expect additional arguments. Usually, service authors want to do that without breaking existing clients.
+
+Candid supports such evolution by defining precise rules that indicate when the new service type will still be able to communicate with all other parties that are using the previous interface description. The underlying formalim is that of _subtyping_.
+
+Services can safely evolve in the following ways:
+
+ * New methods can be added.
+ * Existing methods can return additional values, i.e. the sequence of result types can be extended. Old clients will simply ignore additional values.
+ * Existing methods can shorten their parameter list. Old clients may still send the extra arguments, but they will be ignored.
+ * Existing methods can extend their parameter list with optional arguments (type `opt …`). When reading messages from old clients, who do not pass that argument, a `null` values is assumed.
```

Comment:
Is this safe yet in the higher-order setting? I've lost track of what implemented/intended.

Reply:
No, still not fully sound (see #141). But the intention is sound, we know of one possible fix (subtyping in the decoder)

Reply:
https://en.wikipedia.org/wiki/The_road_to_hell_is_paved_with_good_intentions